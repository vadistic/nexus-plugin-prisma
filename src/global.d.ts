/* need this to build generic/global types*/

export interface PrismaModels {}

export interface PrismaMethods {}

export interface PrismaMappings {}

export interface PrismaOutputs {}

declare global {
  export interface PrismaGen {
    models: PrismaModels
    methods: PrismaMethods
    mappings: PrismaMappings
    outputs: PrismaOutputs
  }
}
