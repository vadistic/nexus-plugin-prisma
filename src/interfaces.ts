import type {
  CommonOutputFieldConfig,
  AllOutputTypes,
  GetGen,
  RootValue,
  ArgsValue,
  MaybePromise,
  ResultValue,
  MaybePromiseDeep,
} from '@nexus/schema/dist/core'
import { GraphQLResolveInfo } from 'graphql'

export type ID = string

export type IDInput = {
  id: ID
}

export type RootName = 'Query' | 'Mutation' | 'Subscription'

export type AllScalarTypes = GetGen<'scalarNames'>
export type AllEnumTypes = GetGen<'enumNames'>
export type AllObjectTypes = GetGen<'objectNames'>
export type AllInterfaceTypes = GetGen<'interfaceNames'>
export type AllUnionTypes = GetGen<'unionNames'>
export type AllObjectLikeTypes = GetGen<'objectNames'> | GetGen<'interfaceNames'>

export type AllPrismaOutputTypes = keyof PrismaGen['outputs'] | keyof PrismaGen['models']

export type PrismaMappings = PrismaGen['mappings']
export type PrismaMethods = PrismaGen['methods']

export type GetTargetMapping<
  TypeName extends string,
  FieldName extends string
> = TypeName extends keyof PrismaMappings
  ? FieldName extends keyof PrismaMappings[TypeName]
    ? PrismaMappings[TypeName][FieldName] extends infer T
      ? T extends keyof PrismaMappings
        ? PrismaMappings[T]
        : never
      : never
    : never
  : never

export type ModelResultValue<TypeName extends string, FieldName extends string> = Pick<
  ResultValue<TypeName, FieldName>,
  keyof GetTargetMapping<TypeName, FieldName>
>

export type PrismaPluginFieldConfig<
  TypeName extends string,
  FieldName extends string
> = CommonOutputFieldConfig<TypeName, FieldName> & {
  type?: AllOutputTypes
  alias?: FieldName

  resolve?: PrismaFieldResolver<TypeName, FieldName>
}

export type PrismaFieldResolver<TypeName extends string, FieldName extends string> = (
  root: RootValue<TypeName>,
  args: ArgsValue<TypeName, FieldName>,
  context: GetGen<'context'>,
  info: GraphQLResolveInfo,
  next: PrismaFieldNext<TypeName, FieldName>,
) =>
  | MaybePromise<ModelResultValue<TypeName, FieldName>>
  | MaybePromiseDeep<ModelResultValue<TypeName, FieldName>>

export type PrismaFieldNext<TypeName extends string, FieldName extends string> = (
  root: RootValue<TypeName>,
  args: ArgsValue<TypeName, FieldName>,
  context: GetGen<'context'>,
  info: GraphQLResolveInfo,
) => ModelResultValue<TypeName, FieldName>

export type PrismaFieldMethod<TypeName extends string> = <FieldName extends string>(
  config?: PrismaPluginFieldConfig<TypeName, FieldName>,
) => void

export type PrismaModelProperty<TypeName extends string> = TypeName extends RootName
  ? never
  : TypeName extends AllPrismaOutputTypes
  ? SimpleModelProperty<TypeName>
  : AliasedModelPluginProperty<TypeName>

export type SimpleModelProperty<TypeName extends keyof PrismaGen['outputs'] & string> = {
  [K in keyof PrismaGen['outputs'][TypeName]]: <FieldName extends string>(
    config?: PrismaPluginFieldConfig<TypeName, FieldName>,
  ) => void
}

export type AliasedModelPluginProperty<TypeName extends string> = <
  ModelName extends AllPrismaOutputTypes
>(
  model: ModelName,
) => {
  [K in keyof PrismaGen['outputs'][ModelName] & string]: PrismaFieldMethod<TypeName>
}

export type PrismaCrudProperty<TypeName extends string> = TypeName extends RootName
  ? PrismaMethods
  : never

export type PrismaCrud<TypeName extends string> = {
  findOne: PrismaFieldMethod<TypeName>
  findMany: PrismaFieldMethod<TypeName>
  create: PrismaFieldMethod<TypeName>
  update: PrismaFieldMethod<TypeName>
  upsert: PrismaFieldMethod<TypeName>
  delete: PrismaFieldMethod<TypeName>
}
