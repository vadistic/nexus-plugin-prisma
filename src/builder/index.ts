export * from './arg'
export * from './model-field'
export * from './enum'
export * from './scalar'
export * from './scalar-config'
export * from './naming'
