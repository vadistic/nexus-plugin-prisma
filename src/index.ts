export * from './plugin'
export * from './config'
export * from './interfaces'
export * from './types'
export * from './builder'
export * from './metadata'
