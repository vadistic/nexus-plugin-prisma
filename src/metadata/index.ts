export * from './dmmf'
export * from './proxy'
export * from './metadata'
