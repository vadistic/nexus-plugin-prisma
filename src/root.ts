export type RootName = 'Query' | 'Mutation' | 'Subscription'

export const rootNames = {
  Query: 'Query',
  Mutation: 'Mutation',
  Subscription: 'Subscription',
} as const

export const rootNameValues: RootName[] = Object.values(rootNames)

export const isRootName = (x: any): x is RootName => rootNameValues.includes(x)
