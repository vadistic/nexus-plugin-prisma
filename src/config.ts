import { PrismaClient } from '@prisma/client'

import { Naming, naming } from './builder/naming'
import { AllScalarTypes } from './interfaces'

export interface PluginPrismaConfig {
  prisma?: PrismaClient
  output?:
    | boolean
    | {
        typegen: string
      }
  naming?: Partial<Naming>
  force?: boolean

  pagination?: {
    takeDefault?: number
    takeLimit?: number
  }

  idType?: 'String' | 'Int' | 'ID' | AllScalarTypes
}

export interface Config {
  prisma: PrismaClient
  output:
    | boolean
    | {
        typegen: string
      }
  naming: Naming
  force: boolean

  pagination: {
    takeDefault: number
    takeLimit: number
  }

  idType: 'String' | 'Int' | 'ID' | AllScalarTypes
}

export const pluginPrismaDefaultConfig: Config = {
  prisma: (undefined as unknown) as PrismaClient, // hacky, but nvm for now
  force: false,
  naming,
  output: false,
  pagination: {
    takeDefault: 25,
    takeLimit: 100,
  },
  idType: 'ID',
}
