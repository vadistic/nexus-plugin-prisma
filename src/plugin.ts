import { plugin, dynamicOutputProperty } from '@nexus/schema'
import { printedGenTypingImport, extendType } from '@nexus/schema/dist/core'
import { PrismaClient } from '@prisma/client'

import { addEnums, addEnumFilters } from './builder/enum'
import { addOutputField } from './builder/model-field'
import { addModelInputs } from './builder/model-input'
import { addScalars, addScalarFilters } from './builder/scalar'
import { PluginPrismaConfig, pluginPrismaDefaultConfig, Config } from './config'
import { getDmmf } from './metadata/dmmf'
import { Metadata } from './metadata/metadata'
import { crudProxy, modelProxy } from './metadata/proxy'
import { typegenPrisma, writeTypegen } from './typegen/typegen'

const MODULE_NAME = '@vadistic/nexus-plugin-prisma'

export const pluginPrisma = (prismaPluginConfig: PluginPrismaConfig) => {
  const config = {
    ...pluginPrismaDefaultConfig,
    ...prismaPluginConfig,
    prisma: prismaPluginConfig.prisma ?? new PrismaClient(),
    naming: {
      ...pluginPrismaDefaultConfig.naming,
      ...prismaPluginConfig.naming,
    },
    pagination: {
      ...pluginPrismaDefaultConfig.pagination,
      ...prismaPluginConfig.pagination,
    },
  } as Config

  const dmmf = getDmmf(config.prisma)
  const metadata = new Metadata(dmmf)

  return plugin({
    name: 'Prisma',
    fieldDefTypes: [
      printedGenTypingImport({
        module: MODULE_NAME,
        bindings: ['PrismaModelProperty', 'PrismaCrudProperty'],
      }),
    ],
    onInstall() {
      return {
        types: [
          dynamicOutputProperty({
            name: 'model',
            typeDefinition: ': PrismaModelProperty<TypeName>',
            factory: ({ typeName, typeDef, stage }) => {
              return modelProxy({
                metadata,
                typeName,
                stage,
                callback: fieldMeta => {
                  addOutputField(config, fieldMeta, typeDef)
                },
              })
            },
          }),
          dynamicOutputProperty({
            name: 'crud',
            typeDefinition: ': PrismaCrudProperty<TypeName>',
            factory: ({ typeName, typeDef, stage }) => {
              return crudProxy({
                metadata,
                typeName,
                stage,
                callback: fieldMeta => {
                  addOutputField(config, fieldMeta, typeDef)
                },
              })
            },
          }),
        ],
      }
    },

    onBeforeBuild: lens => {
      addScalars(config, metadata, lens)
      addScalarFilters(config, metadata, lens)

      addEnums(config, metadata, lens)
      addEnumFilters(config, metadata, lens)

      addModelInputs(config, metadata, lens)

      metadata.types.forEach(typeMeta => {
        if (!typeMeta.root) return

        const extension = extendType<any>({
          type: typeMeta.typeName,
          definition(t) {
            typeMeta.fields.forEach(fieldMeta => {
              if (fieldMeta.outputField.outputType.kind !== 'object') return

              addOutputField(config, fieldMeta, t, true)
            })
          },
        })

        lens.addType(extension)
      })
    },

    onAfterBuild() {
      writeTypegen(
        config,
        typegenPrisma(
          [
            { module: '@prisma/client', default: '* as Prisma' },
            { module: MODULE_NAME, bindings: ['PrismaCrud'] },
          ],
          metadata,
        ),
      )
    },
  })
}
