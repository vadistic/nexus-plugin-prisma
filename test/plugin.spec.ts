import { objectType, makeSchema } from '@nexus/schema'
import { printSchema } from 'graphql'

import { pluginPrisma } from '../src/plugin'

describe('plugin', () => {
  const TagModel = objectType({
    name: 'Tag',
    definition(t: any) {
      t.model.id()
    },
  })

  const schema = makeSchema({
    types: [TagModel],
    plugins: [pluginPrisma({ output: false })],
    outputs: false,
  })

  const schemaForce = makeSchema({
    types: [TagModel],
    plugins: [pluginPrisma({ output: false, force: true })],
    outputs: false,
  })

  test('ok', () => {
    expect(schema).toBeDefined()
    expect(schemaForce).toBeDefined()
  })

  test('print force schema', () => {
    expect(printSchema(schemaForce)).toMatchSnapshot('schema.graphql')
  })

  test('print schema', () => {
    expect(printSchema(schema)).toMatchInlineSnapshot(`
      "type Tag {
        id: String!
      }

      type Query {
        ok: Boolean!
      }
      "
    `)
  })
})
