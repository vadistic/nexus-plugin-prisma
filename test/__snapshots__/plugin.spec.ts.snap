// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin print force schema: schema.graphql 1`] = `
"type Tag {
  id: String!
}

\\"\\"\\"\\"\\"\\"
scalar Json

\\"\\"\\"\\"\\"\\"
scalar DateTime

input JsonNullableFilter {
  \\"\\"\\"\\"\\"\\"
  equals: Json

  \\"\\"\\"\\"\\"\\"
  not: Json

  \\"\\"\\"\\"\\"\\"
  in: [Json]

  \\"\\"\\"\\"\\"\\"
  notInt: [Json]

  \\"\\"\\"\\"\\"\\"
  exists: Boolean
}

input JsonFilter {
  \\"\\"\\"\\"\\"\\"
  equals: Json

  \\"\\"\\"\\"\\"\\"
  not: Json

  \\"\\"\\"\\"\\"\\"
  in: [Json]

  \\"\\"\\"\\"\\"\\"
  notInt: [Json]
}

input StringNullableFilter {
  \\"\\"\\"\\"\\"\\"
  equals: String

  \\"\\"\\"\\"\\"\\"
  not: String

  \\"\\"\\"\\"\\"\\"
  startsWith: String

  \\"\\"\\"\\"\\"\\"
  endsWith: String

  \\"\\"\\"\\"\\"\\"
  contains: String

  \\"\\"\\"\\"\\"\\"
  in: [String]

  \\"\\"\\"\\"\\"\\"
  notInt: [String]

  \\"\\"\\"\\"\\"\\"
  exists: Boolean
}

input StringFilter {
  \\"\\"\\"\\"\\"\\"
  equals: String

  \\"\\"\\"\\"\\"\\"
  not: String

  \\"\\"\\"\\"\\"\\"
  startsWith: String

  \\"\\"\\"\\"\\"\\"
  endsWith: String

  \\"\\"\\"\\"\\"\\"
  contains: String

  \\"\\"\\"\\"\\"\\"
  in: [String]

  \\"\\"\\"\\"\\"\\"
  notInt: [String]
}

input IntNullableFilter {
  \\"\\"\\"\\"\\"\\"
  equals: Int

  \\"\\"\\"\\"\\"\\"
  not: Int

  \\"\\"\\"\\"\\"\\"
  lt: Int

  \\"\\"\\"\\"\\"\\"
  lte: Int

  \\"\\"\\"\\"\\"\\"
  gt: Int

  \\"\\"\\"\\"\\"\\"
  gte: Int

  \\"\\"\\"\\"\\"\\"
  in: [Int]

  \\"\\"\\"\\"\\"\\"
  notInt: [Int]

  \\"\\"\\"\\"\\"\\"
  exists: Boolean
}

input IntFilter {
  \\"\\"\\"\\"\\"\\"
  equals: Int

  \\"\\"\\"\\"\\"\\"
  not: Int

  \\"\\"\\"\\"\\"\\"
  lt: Int

  \\"\\"\\"\\"\\"\\"
  lte: Int

  \\"\\"\\"\\"\\"\\"
  gt: Int

  \\"\\"\\"\\"\\"\\"
  gte: Int

  \\"\\"\\"\\"\\"\\"
  in: [Int]

  \\"\\"\\"\\"\\"\\"
  notInt: [Int]
}

input FloatNullableFilter {
  \\"\\"\\"\\"\\"\\"
  equals: Float

  \\"\\"\\"\\"\\"\\"
  not: Float

  \\"\\"\\"\\"\\"\\"
  lt: Float

  \\"\\"\\"\\"\\"\\"
  lte: Float

  \\"\\"\\"\\"\\"\\"
  gt: Float

  \\"\\"\\"\\"\\"\\"
  gte: Float

  \\"\\"\\"\\"\\"\\"
  in: [Float]

  \\"\\"\\"\\"\\"\\"
  notInt: [Float]

  \\"\\"\\"\\"\\"\\"
  exists: Boolean
}

input FloatFilter {
  \\"\\"\\"\\"\\"\\"
  equals: Float

  \\"\\"\\"\\"\\"\\"
  not: Float

  \\"\\"\\"\\"\\"\\"
  lt: Float

  \\"\\"\\"\\"\\"\\"
  lte: Float

  \\"\\"\\"\\"\\"\\"
  gt: Float

  \\"\\"\\"\\"\\"\\"
  gte: Float

  \\"\\"\\"\\"\\"\\"
  in: [Float]

  \\"\\"\\"\\"\\"\\"
  notInt: [Float]
}

input DateTimeNullableFilter {
  \\"\\"\\"\\"\\"\\"
  equals: DateTime

  \\"\\"\\"\\"\\"\\"
  not: DateTime

  \\"\\"\\"\\"\\"\\"
  lt: DateTime

  \\"\\"\\"\\"\\"\\"
  lte: DateTime

  \\"\\"\\"\\"\\"\\"
  gt: DateTime

  \\"\\"\\"\\"\\"\\"
  gte: DateTime

  \\"\\"\\"\\"\\"\\"
  in: [DateTime]

  \\"\\"\\"\\"\\"\\"
  notInt: [DateTime]

  \\"\\"\\"\\"\\"\\"
  exists: Boolean
}

input DateTimeFilter {
  \\"\\"\\"\\"\\"\\"
  equals: DateTime

  \\"\\"\\"\\"\\"\\"
  not: DateTime

  \\"\\"\\"\\"\\"\\"
  lt: DateTime

  \\"\\"\\"\\"\\"\\"
  lte: DateTime

  \\"\\"\\"\\"\\"\\"
  gt: DateTime

  \\"\\"\\"\\"\\"\\"
  gte: DateTime

  \\"\\"\\"\\"\\"\\"
  in: [DateTime]

  \\"\\"\\"\\"\\"\\"
  notInt: [DateTime]
}

input BooleanNullableFilter {
  \\"\\"\\"\\"\\"\\"
  equals: Boolean

  \\"\\"\\"\\"\\"\\"
  not: Boolean

  \\"\\"\\"\\"\\"\\"
  in: [Boolean]

  \\"\\"\\"\\"\\"\\"
  notInt: [Boolean]

  \\"\\"\\"\\"\\"\\"
  exists: Boolean
}

input BooleanFilter {
  \\"\\"\\"\\"\\"\\"
  equals: Boolean

  \\"\\"\\"\\"\\"\\"
  not: Boolean

  \\"\\"\\"\\"\\"\\"
  in: [Boolean]

  \\"\\"\\"\\"\\"\\"
  notInt: [Boolean]
}

input IDNullableFilter {
  \\"\\"\\"\\"\\"\\"
  equals: ID

  \\"\\"\\"\\"\\"\\"
  not: ID

  \\"\\"\\"\\"\\"\\"
  in: [ID]

  \\"\\"\\"\\"\\"\\"
  notInt: [ID]

  \\"\\"\\"\\"\\"\\"
  exists: Boolean
}

input IDFilter {
  \\"\\"\\"\\"\\"\\"
  equals: ID

  \\"\\"\\"\\"\\"\\"
  not: ID

  \\"\\"\\"\\"\\"\\"
  in: [ID]

  \\"\\"\\"\\"\\"\\"
  notInt: [ID]
}

enum SortDirection {
  ASC
  DESC
}

enum WorkspaceType {
  FREE
}

enum StageType {
  INITIAL
  PROCESS
  REJECT
  SUCCESS
}

enum FieldType {
  TEXT
  TEXTAREA
  NUMBER
  SCORE
  BOOLEAN
}

enum JobStatus {
  DRAFT
  OPEN
  CLOSED
}

enum TagType {
  CANDIDATE
  APPLICATION
  JOB
}

input WorkspaceTypeNullableEnumFilter {
  \\"\\"\\"\\"\\"\\"
  equals: WorkspaceType

  \\"\\"\\"\\"\\"\\"
  not: WorkspaceType

  \\"\\"\\"\\"\\"\\"
  in: [WorkspaceType]

  \\"\\"\\"\\"\\"\\"
  notInt: [WorkspaceType]

  \\"\\"\\"\\"\\"\\"
  exists: Boolean
}

input WorkspaceTypeEnumFilter {
  \\"\\"\\"\\"\\"\\"
  equals: WorkspaceType

  \\"\\"\\"\\"\\"\\"
  not: WorkspaceType

  \\"\\"\\"\\"\\"\\"
  in: [WorkspaceType]

  \\"\\"\\"\\"\\"\\"
  notInt: [WorkspaceType]
}

input StageTypeNullableEnumFilter {
  \\"\\"\\"\\"\\"\\"
  equals: StageType

  \\"\\"\\"\\"\\"\\"
  not: StageType

  \\"\\"\\"\\"\\"\\"
  in: [StageType]

  \\"\\"\\"\\"\\"\\"
  notInt: [StageType]

  \\"\\"\\"\\"\\"\\"
  exists: Boolean
}

input StageTypeEnumFilter {
  \\"\\"\\"\\"\\"\\"
  equals: StageType

  \\"\\"\\"\\"\\"\\"
  not: StageType

  \\"\\"\\"\\"\\"\\"
  in: [StageType]

  \\"\\"\\"\\"\\"\\"
  notInt: [StageType]
}

input FieldTypeNullableEnumFilter {
  \\"\\"\\"\\"\\"\\"
  equals: FieldType

  \\"\\"\\"\\"\\"\\"
  not: FieldType

  \\"\\"\\"\\"\\"\\"
  in: [FieldType]

  \\"\\"\\"\\"\\"\\"
  notInt: [FieldType]

  \\"\\"\\"\\"\\"\\"
  exists: Boolean
}

input FieldTypeEnumFilter {
  \\"\\"\\"\\"\\"\\"
  equals: FieldType

  \\"\\"\\"\\"\\"\\"
  not: FieldType

  \\"\\"\\"\\"\\"\\"
  in: [FieldType]

  \\"\\"\\"\\"\\"\\"
  notInt: [FieldType]
}

input JobStatusNullableEnumFilter {
  \\"\\"\\"\\"\\"\\"
  equals: JobStatus

  \\"\\"\\"\\"\\"\\"
  not: JobStatus

  \\"\\"\\"\\"\\"\\"
  in: [JobStatus]

  \\"\\"\\"\\"\\"\\"
  notInt: [JobStatus]

  \\"\\"\\"\\"\\"\\"
  exists: Boolean
}

input JobStatusEnumFilter {
  \\"\\"\\"\\"\\"\\"
  equals: JobStatus

  \\"\\"\\"\\"\\"\\"
  not: JobStatus

  \\"\\"\\"\\"\\"\\"
  in: [JobStatus]

  \\"\\"\\"\\"\\"\\"
  notInt: [JobStatus]
}

input TagTypeNullableEnumFilter {
  \\"\\"\\"\\"\\"\\"
  equals: TagType

  \\"\\"\\"\\"\\"\\"
  not: TagType

  \\"\\"\\"\\"\\"\\"
  in: [TagType]

  \\"\\"\\"\\"\\"\\"
  notInt: [TagType]

  \\"\\"\\"\\"\\"\\"
  exists: Boolean
}

input TagTypeEnumFilter {
  \\"\\"\\"\\"\\"\\"
  equals: TagType

  \\"\\"\\"\\"\\"\\"
  not: TagType

  \\"\\"\\"\\"\\"\\"
  in: [TagType]

  \\"\\"\\"\\"\\"\\"
  notInt: [TagType]
}

input TagWhere {
  \\"\\"\\"Filter by Tag's id\\"\\"\\"
  id: StringFilter

  \\"\\"\\"\\"\\"\\"
  AND: [TagWhere]

  \\"\\"\\"\\"\\"\\"
  OR: [TagWhere]

  \\"\\"\\"\\"\\"\\"
  NOT: TagWhere
}

input TagFilter {
  \\"\\"\\"\\"\\"\\"
  every: TagWhere

  \\"\\"\\"\\"\\"\\"
  none: TagWhere

  \\"\\"\\"\\"\\"\\"
  some: TagWhere
}

input TagOrderBy {
  id: SortDirection
}

type Query {
  ok: Boolean!
}
"
`;
